File: leetcode/cpp/133.cpp
#include <bits/stdc++.h>
using namespace std;

// Definition for a Node.
class Node
{
public:
    int val;
    vector<Node *> neighbors;
    Node()
    {
        val = 0;
        neighbors = vector<Node *>();
    }
    Node(int _val)
    {
        val = _val;
        neighbors = vector<Node *>();
    }
    Node(int _val, vector<Node *> _neighbors)
    {
        val = _val;
        neighbors = _neighbors;
    }
};

class Solution
{
public:
    Node *cloneGraph(Node *node)
    {
        if (node == nullptr)
        {
            return nullptr;
        }
        map<int, Node *> clonedNodes;
        // BFS
        queue<Node *> q;
        q.push(node);
        clonedNodes[node->val] = new Node(node->val);
        while (q.size() > 0)
        {
            Node *current = q.front();
            q.pop();
            for (auto neighbor : current->neighbors)
            {
                if (clonedNodes.find(neighbor->val) == clonedNodes.end())
                {
                    clonedNodes[neighbor->val] = new Node(neighbor->val);
                    q.push(neighbor);
                }
                clonedNodes[current->val]->neighbors.push_back(clonedNodes[neighbor->val]);
            }
        }
        return clonedNodes[node->val];
    }
};

File: leetcode/python/133.py
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []

from typing import Optional, DefaultDict, Deque
class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        if node == None:
            return None
        clonedNodes =  DefaultDict(Node)
        clonedNodes[node.val] = Node(node.val,node.neighbors)
        q = []
        while len(q) > 0:
            current = q[0]
            q = q[1:]
            for neighbor in current.neighbors:
                clonedNodes[neighbor.val] = Node(neighbor.val,neighbor.neighbors)
                q.append(neighbor)
        return clonedNodes[node.val]